include/fiction/algorithms/simulation/sidb/quickexact.hpp

```diff
@@ -0,0 +1,506 @@
+                                                    // automatically changed based on the new charge distribution.
+        }
+
+        // Charge configurations of the sub-layout are iterated
+        while (charge_layout.get_charge_index_sub_layout().first < charge_layout.get_max_charge_index_sub_layout())
+        {
+            if (charge_layout.is_physically_valid())
+            {
+                charge_distribution_surface<Lyt> charge_lyt_copy{charge_layout};
+                charge_lyt_copy.recompute_system_energy();
+
+                // The previously detected negatively-charged SiDBs are added to the final layout.
+                for (const auto& cell : detected_negative_sidbs)
+                {
+                    charge_lyt_copy.adding_sidb_to_layout(cell, -1);
+                }
+
+                result.charge_distributions.push_back(charge_lyt_copy);
+            }
+
+            charge_layout.increase_charge_index_of_sub_layout_by_one(
+                dependent_cell_mode::VARIABLE, energy_calculation::KEEP_OLD_ENERGY_VALUE,
+                charge_distribution_history::CONSIDER, exhaustive_algorithm::QUICKEXACT);
+        }
+
+        if (charge_layout.is_physically_valid())
+        {
+            charge_distribution_surface<Lyt> charge_lyt_copy{charge_layout};
+
+            for (const auto& cell : detected_negative_sidbs)
+            {
+                charge_lyt_copy.adding_sidb_to_layout(cell, -1);
+            }
+
+            result.charge_distributions.push_back(charge_lyt_copy);
+        }
+
+        for (const auto& cell : detected_negative_sidbs)
+        {
+            layout.assign_cell_type(cell, Lyt::cell_type::NORMAL);
+        }
+    }
+    /**
+     * This function is required for the initialization.
+     */
+    void initialize_charge_layout()
+    {
+        if constexpr (has_get_sidb_defect_v<Lyt>)
+        {
+            layout.foreach_sidb_defect(
+                [this](const auto& cd)
+                {
+                    const auto& [cell, defect] = cd;
+
+                    if (defect.type != sidb_defect_type::NONE)
+                    {
+                        charge_lyt.add_defect(cell, layout.get_sidb_defect(cell));
+                    }
+                });
+        }
+
+        charge_lyt.assign_local_external_potential(params.local_external_potential);
+        charge_lyt.assign_global_external_potential(params.global_potential, dependent_cell_mode::VARIABLE);
+
+        detected_negative_sidb_indices = charge_lyt.negative_sidb_detection();
+        detected_negative_sidbs.reserve(detected_negative_sidb_indices.size());
+
+        all_sidbs_in_lyt_without_negative_detected_ones = charge_lyt.get_sidb_order();
+        real_placed_defects                             = charge_lyt.get_defects();
+        number_of_sidbs                                 = charge_lyt.get_sidb_order().size();
```

Comment:
Why not `charge_lyt.num_cells()`? Isn't it better to simply get a number returned rather than first make a copy of all cells and then fetch their number?